src/RobotPool.cpp

```diff
@@ -143,23 +143,22 @@ bool RobotPool::robotAvailable(RobotType _type)
 	switch (_type)
 	{
 	case ROBOT_DIGGER:
+	{
 		return getDigger() != nullptr;
-		break;
-
+	}
 	case ROBOT_DOZER:
+	{
 		return getDozer() != nullptr;
-		break;
-
+	}
 	case ROBOT_MINER:
+	{
 		return getMiner() != nullptr;
-		break;
-
+	}
 	default:
+	{
 		return false;
-		break;
 	}
-
-	return false;
+	}
```

Comment:
Minor musing. I'm wondering if this might be better removing the `default` case, and putting the `return false;` outside of the switch. Not sure which would be better.

Initially I missed seeing the `default` case and wondered if this might not return a value in some cases. In that sense, having it outside the `switch` made the return result more obvious. On the other hand, using `default` might say something about the cases which are specifically handled.

I rarely use `switch`, so I don't really have any guidelines on making such a decision.

Reply:
Conventional guidelines are to A) Make a `case` for every value of your switch variable if possible (duplicates not withstanding, but document it; and B) Always provide a `default` case.

In fact, Visual Assist even has a refactor option "Add missing case statements" for uses with `enum` and `enum class` types that does both A and B. If you are missing a statement it adds it, and it adds the `default` case if it is missing.

By providing a `default` any value not specified in the `case` labels will jump to the default. In this case if the value is not one of `ROBOT_DOZER`, `ROBOT_DIGGER`, `ROBOT_MINER`, `false` will be returned.

Additionally, I have a personal rule that if the range of your switch variable can ever change, loudly announce with a `MessageWindow` and a fatal error/`exit(0)` in the default case it when it does without your knowledge.

In this case, the number of robot types could change (I remember there were airbots in the original manual and research tree, but they were never implemented in-game.) and there should be a loud failure that the number of the type of robots has changed and they aren't handled in a `switch` case.

See
https://github.com/cugone/Abrams2019/blob/870dcd04d6a8fa8e5132efb015126d0de6935ec0/Engine/Code/Engine/Core/KerningFont.cpp#L858
for an example of announcing that a third-party API has changed and the new value isn't being handled; and
https://github.com/cugone/Abrams2019/blob/870dcd04d6a8fa8e5132efb015126d0de6935ec0/Engine/Code/Engine/Renderer/DirectX/DX11.cpp#L673
for using sensible defaults when needed and failing when the default is reached due to required state not being met.

Reply:
I think I would prefer an exception for a `MessageWindow` or `exit(0)`, though yes, it should loudly announce when not all cases are handled. I believe D added `final` to `switch` statements as a way of enforcing all cases are handled, and would result in a compiler error if they weren't.